### PR TITLE
issue79 reformat the code

### LINKED ISSUE #79: make .py files pep conform
Atm not all files are pep conform.
It would be best practise if we would reformat the code.

### PR DESCRIPTION
**Pull request by rawbby**
resolves #79

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -4,7 +4,7 @@
         :target: https://travis-ci.com/lettucecfd/lettuce
 
 .. .. image:: https://img.shields.io/pypi/v/lettuce.svg
-        :target: https://pypi.python.org/pypi/lettuce
+:target: https://pypi.python.org/pypi/lettuce
```
Comment (rawbby) [inline README.rst:7]: This looks like an error to me :/
Should I remove the rst-file changes entirely from the PR or is there another issue in this lines?

File: tests/test_boundaries.py [hunk 2]
```diff
@@ -68,25 +67,48 @@ def test_anti_bounce_back_outlet(f_lattice):
         if Q == 27:
             u_w = u[:, -1, :, :] + 0.5 * (u[:, -1, :, :] - u[:, -2, :, :])
             for i in [1, 11, 13, 15, 17, 19, 21, 23, 25]:
-                f_ref[lattice.stencil.opposite[i], -1, :, :] = - f_ref[i, -1, :, :] + lattice.stencil.w[i] * lattice.rho(f)[0, -1, :, :] * \
-                    (2 + torch.einsum('c, cyz -> yz', torch.tensor(lattice.stencil.e[i], device=f.device, dtype=f.dtype), u_w) ** 2 / lattice.stencil.cs ** 4 - (torch.norm(u_w, dim=0) / lattice.stencil.cs) ** 2)
+                f_ref[lattice.stencil.opposite[i], -1, :, :] = - f_ref[i, -1, :, :] + lattice.stencil.w[
```
Comment (rawbby) [inline tests/test_boundaries.py:70]: Maybe there is a way that is not that ugly but also satisfies pep...

**Review by rawbby — COMMENTED** [review]
I will try to change some section by hand to make the code "not ugly" and pep conform.

File: README.rst (hunk 1 above)
Comment (Olllom) [reply README.rst:7]: this is an error, yes. You can just remove the line altogether; this comes from cookiecutter. The code is not yet hosted on pypi and even then `lettuce` is surely taken.

File: tests/test_boundaries.py (hunk 2 above)
Comment (Olllom) [reply tests/test_boundaries.py:70]: Yes, please ;-)
Maybe just open parentheses behind the `=` and break lines for each summand.

File: tests/test_boundaries.py [hunk 3]
```diff
@@ -40,53 +39,83 @@ def test_equilibrium_boundary_pu(f_lattice):
...
         direction = [1, 0, 0]
+
         if Q == 27:
             u_w = u[:, -1, :, :] + 0.5 * (u[:, -1, :, :] - u[:, -2, :, :])
+            u_w_norm = torch.norm(u_w, dim=0)
+
             for i in [1, 11, 13, 15, 17, 19, 21, 23, 25]:
-                f_ref[lattice.stencil.opposite[i], -1, :, :] = - f_ref[i, -1, :, :] + lattice.stencil.w[i] * lattice.rho(f)[0, -1, :, :] * \
-                    (2 + torch.einsum('c, cyz -> yz', torch.tensor(lattice.stencil.e[i], device=f.device, dtype=f.dtype), u_w) ** 2 / lattice.stencil.cs ** 4 - (torch.norm(u_w, dim=0) / lattice.stencil.cs) ** 2)
+                stencil_e_tensor = torch.tensor(lattice.stencil.e[i], device=f.device, dtype=f.dtype)
```
Comment (rawbby) [inline tests/test_boundaries.py:74]: I think this format is acceptable.
Can someone confirm that I have not done any typos in this file?

**Comment by rawbby** [comment]
From my side that's it. Please comment if you like me to further tidy something specific.

File: lettuce/boundary.py [hunk 4]
```diff
@@ -68,13 +69,24 @@ class AntiBounceBackOutlet:
...
+
+        assert len(direction) in [1, 2, 3], \
+            LettuceException(
+                f"Invalid direction parameter. Expected direction of of length 1, 2 or 3 but got {len(direction)}.")
+
+        assert (direction.count(0) == (len(direction) - 1)) and ((1 in direction) ^ (-1 in direction)), \
+            LettuceException(
+                "Invalid direction parameter. "
+                f"Expected direction with all entries 0 except one 1 or -1 but got {direction}.")
+
```
Comment (Olllom) [inline lettuce/boundary.py:85]: That's much better but still quite ugly. I actually don't think these lines will ever throw a LettuceException. Rather, they will raise AssertionError which does not seem to be the desired behavior. But we can also fix this later after this PR is merged.

File: lettuce/cli.py [hunk 5]
```diff
@@ -91,48 +91,49 @@ def benchmark(ctx, steps, resolution, profile_out, flow, vtk_out):
...
 
         # Simulation
-        flow = TaylorGreenVortex2D(resolution=resolution, reynolds_number=10000, mach_number=mach_number, lattice=lattice)
+        flow = TaylorGreenVortex2D(resolution=resolution, reynolds_number=10000, mach_number=mach_number,
+                                   lattice=lattice)
         collision = BGKCollision(lattice, tau=flow.units.relaxation_parameter_lu)
         streaming = StandardStreaming(lattice)
         simulation = Simulation(flow=flow, lattice=lattice, collision=collision, streaming=streaming)
-        if(init_f_neq):
+        if (init_f_neq):
```
Comment (Olllom) [inline lettuce/cli.py:119]: These parentheses are not needed.

File: lettuce/flows/obstacle.py [hunk 6]
```diff
@@ -71,8 +72,12 @@ def initial_solution(self, x):
 
     @property
     def grid(self):
-        x = np.linspace(0, self.resolution_x / self.units.characteristic_length_lu, num=self.resolution_x, endpoint=False)
-        y = np.linspace(0, self.resolution_y / self.units.characteristic_length_lu, num=self.resolution_y, endpoint=False)
+        stop_x = self.resolution_x / self.units.characteristic_length_lu
+        stop_y = self.resolution_x / self.units.characteristic_length_lu
```
Comment (Olllom) [inline lettuce/flows/obstacle.py:76]: This is introducing a bug. Should be resolution_y.

File: lettuce/flows/obstacle.py [hunk 7]
```diff
@@ -121,9 +127,14 @@ def initial_solution(self, x):
 
     @property
     def grid(self):
-        x = np.linspace(0, self.resolution_x / self.units.characteristic_length_lu, num=self.resolution_x, endpoint=False)
-        y = np.linspace(0, self.resolution_y / self.units.characteristic_length_lu, num=self.resolution_y, endpoint=False)
-        z = np.linspace(0, self.resolution_z / self.units.characteristic_length_lu, num=self.resolution_z, endpoint=False)
+        stop_x = self.resolution_x / self.units.characteristic_length_lu
+        stop_y = self.resolution_x / self.units.characteristic_length_lu
```
Comment (Olllom) [inline lettuce/flows/obstacle.py:131]: See above

File: lettuce/lattices.py [hunk 8]
```diff
@@ -44,7 +43,8 @@ def Q(self):
     def convert_to_tensor(self, array):
         with warnings.catch_warnings():
             warnings.simplefilter("ignore")
-            if isinstance(array, np.ndarray) and array.dtype in [np.bool, np.uint8] or isinstance(array, torch.BoolTensor):
+            if isinstance(array, np.ndarray)\
+                    and array.dtype in [np.bool, np.uint8] or isinstance(array, torch.BoolTensor):
```
Comment (Olllom) [inline lettuce/lattices.py:47]: Urgh. Can we replace the backslash notation by parentheses?

File: tests/test_unit.py [hunk 9]
```diff
@@ -11,32 +11,35 @@
...
+                           characteristic_density_pu=0.7)
     assert units.convert_velocity_to_lu(units.convert_velocity_to_pu(2.0)) == pytest.approx(2.0)
     assert units.convert_time_to_lu(units.convert_time_to_pu(2.0)) == pytest.approx(2.0)
     assert units.convert_length_to_lu(units.convert_length_to_pu(2.0)) == pytest.approx(2.0)
     assert units.convert_density_to_lu(units.convert_density_to_pu(2.0)) == pytest.approx(2.0)
     assert units.convert_pressure_to_lu(units.convert_pressure_to_pu(2.0)) == pytest.approx(2.0)
     assert units.convert_density_lu_to_pressure_pu(units.convert_pressure_pu_to_density_lu(2.0)) == pytest.approx(2.0)
     assert units.convert_energy_to_lu(units.convert_energy_to_pu(2.0)) == pytest.approx(2.0)
-    assert units.convert_incompressible_energy_to_lu(units.convert_incompressible_energy_to_pu(2.0)) == pytest.approx(2.0)
+    assert units.convert_incompressible_energy_to_lu(units.convert_incompressible_energy_to_pu(2.0)) == pytest.approx(
```
Comment (Olllom) [inline tests/test_unit.py:31]: Urgh

File: tests/test_simulation.py [hunk 10]
```diff
@@ -88,4 +88,4 @@ def test_initialize_fneq(Case, dtype_device):
         error_u, error_p = np.mean(np.abs(error_reporter_neq.out), axis=0).tolist()
         error_u_eq, error_p_eq = np.mean(np.abs(error_reporter_eq.out), axis=0).tolist()
 
-        assert(error_u < error_u_eq)
+        assert (error_u < error_u_eq)
```
Comment (Olllom) [inline tests/test_simulation.py:91]: Redundant parentheses

File: tests/test_reporters.py [hunk 11]
```diff
@@ -34,9 +35,9 @@ def test_generic_reporters(Observable, Case, dtype_device):
     simulation.step(2)
     values = np.asarray(reporter.out)
     if Observable is EnergySpectrum:
-        assert (values[1, 2:] == pytest.approx(values[0, 2:], rel=0.0, abs=values[0, 2:].sum()/10))
+        assert (values[1, 2:] == pytest.approx(values[0, 2:], rel=0.0, abs=values[0, 2:].sum() / 10))
     else:
-        assert(values[1, 2] == pytest.approx(values[0, 2], rel=0.05))
+        assert (values[1, 2] == pytest.approx(values[0, 2], rel=0.05))
```
Comment (Olllom) [inline tests/test_reporters.py:40]: No parentheses

File: tests/test_reporters.py [hunk 12]
```diff
@@ -34,9 +35,9 @@ def test_generic_reporters(Observable, Case, dtype_device):
     simulation.step(2)
     values = np.asarray(reporter.out)
     if Observable is EnergySpectrum:
-        assert (values[1, 2:] == pytest.approx(values[0, 2:], rel=0.0, abs=values[0, 2:].sum()/10))
+        assert (values[1, 2:] == pytest.approx(values[0, 2:], rel=0.0, abs=values[0, 2:].sum() / 10))
```
Comment (Olllom) [inline tests/test_reporters.py:38]: No parentheses

File: lettuce/simulation.py [hunk 13]
```diff
@@ -134,12 +135,13 @@ def initialize_f_neq(self):
             grad_u2 = torch_gradient(u[2], dx=1, order=6)[None, ...]
             S = torch.cat([S, grad_u2])
 
-        Pi_1 = 1.0 * self.flow.units.relaxation_parameter_lu * rho * S / self.lattice.cs**2
-        Q = torch.einsum('ia,ib->iab', [self.lattice.e, self.lattice.e]) - torch.eye(self.lattice.D, device=self.lattice.device, dtype=self.lattice.dtype)*self.lattice.cs**2
+        Pi_1 = 1.0 * self.flow.units.relaxation_parameter_lu * rho * S / self.lattice.cs ** 2
+        Q = torch.einsum('ia,ib->iab', [self.lattice.e, self.lattice.e]) \
+            - torch.eye(self.lattice.D, device=self.lattice.device, dtype=self.lattice.dtype) * self.lattice.cs ** 2
```
Comment (Olllom) [inline lettuce/simulation.py:140]: Can we solve this without backslash?

**Review by Olllom — CHANGES_REQUESTED** [review]
LGTM mostly. See Details in line.

File: lettuce/boundary.py (hunk 4 above)
Comment (rawbby) [reply lettuce/boundary.py:85]: I made an issue for this

File: tests/test_boundaries.py (hunk 3 above)
Comment (Olllom) [reply tests/test_boundaries.py:74]: lgtm

**Comment by Olllom** [comment]
Excellent @rawbby, thanks a ton. I'll merge this.

**Review by Olllom — APPROVED** [review]

**Pull request merged**